sarc/core/scraping/users.py

```diff
@@ -0,0 +1,187 @@
+    google_scholar_profile: ValidField[str] = Field(default_factory=ValidField[str])
+
+    # This is not really required for serialization, but it makes the order or
+    # known_matches random if not present and complicates testing. If it becomes
+    # a problem outside of tests we can find another solution.
+    @field_serializer("known_matches", when_used="json")
+    def _serialize_deterministic(self, value: set[MatchID]):
+        return sorted(value, key=lambda m: (m.name, m.mid))
+
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, UserMatch) and self.matching_id == other.matching_id
+
+    def __hash__(self) -> int:
+        return hash(self.matching_id)
+
+
+# plugins are run in the order they are defined in the config file and the first plugin to define a value wins.
+class UserScraper[T](Protocol):
+    config_type: Type[T]
+
+    def validate_config(self, config_data: Any) -> T:
+        return deserialize(self.config_type, config_data)
+
+    def get_user_data(self, config: T) -> bytes: ...  # pragma: nocover
+
+    def parse_user_data(
+        self, config: T, data: bytes
+    ) -> Iterable[UserMatch]: ...  # pragma: nocover
+
+
+_builtin_scrapers: dict[str, UserScraper] = dict()
+_user_scrapers = entry_points(group="sarc.user_scraper")
+
+
+def get_user_scraper(name: str) -> UserScraper:
+    """Raises KeyError if the name is not found"""
+    try:
+        return _builtin_scrapers[name]
+    except KeyError:
+        pass
+    val = _user_scrapers[name]
+    return val.load()()
+
+
+def update_user_match(*, value: UserMatch, update: UserMatch) -> None:
+    """
+    Fills in any missing information in value with the data in update.
+    """
+    if value.display_name is None:
+        value.display_name = update.display_name
+
+    if value.email is None:
+        value.email = update.email
+
+    name_dict = {mid.name: mid for mid in value.known_matches}
+
+    # Add the matching ids of the new usermatch to make sure that we have all
+    # the ids that this user is known under.
+    assert (
+        name_dict.get(update.matching_id.name, update.matching_id) == update.matching_id
+    )
+    value.known_matches.add(update.matching_id)
+    for mid in update.known_matches:
+        assert name_dict.get(mid.name, mid) == mid
+        value.known_matches.add(mid)
+
+    for domain, credentials in update.associated_accounts.items():
+        if domain not in value.associated_accounts:
+            value.associated_accounts[domain] = credentials
+        else:
+            value.associated_accounts[domain].merge_with(credentials, truncate=True)
+
+    value.supervisor.merge_with(update.supervisor, truncate=True)
+    value.co_supervisors.merge_with(update.co_supervisors, truncate=True)
+
+    value.github_username.merge_with(update.github_username, truncate=True)
+    value.google_scholar_profile.merge_with(
+        update.google_scholar_profile, truncate=True
+    )
```

Comment:
Should we also update `value.member_type` with `update.member_type` ?

Reply:
Yes, done